### PR TITLE
Add support for > 32 qudits to cirq.sample_state_vector. Fix for #6031

### PR DESCRIPTION
**Pull request by NoureldinYosri**
- When the number of qudits > 32 do the transpose operation on the flattened array. 
- I expect this to be slower than `np.transpose(tensor, axes)` so I'm keeping the use of `np.transpose` when we can.
- I also rewrote the `_prob` in a more efficient way. The old way did a lot of extra work when reordering the axes is all that is needed.

Fix for #6031.

### CODE REVIEW

File: cirq-core/cirq/sim/state_vector.py [hunk 1]
```diff
@@ -325,30 +326,24 @@ def measure_state_vector(
...
-                    for b in range(np.prod(meas_shape, dtype=np.int64))
-                ]
-            )
-            ** 2
-        )
-        probs = np.sum(probs, axis=tuple(range(1, len(probs.shape))))
+        # If we can't use numpy due to numpy/numpy#5744, use a slower method.
+        probs = linalg.transpose_flattened_array(probs, qid_shape, list(indices) + not_measured)
+
+    if len(not_measured):
```
Comment (rht) [inline cirq-core/cirq/sim/state_vector.py:341]: This is simpler than the old code. I suppose it is also faster?

File: cirq-core/cirq/sim/state_vector.py [hunk 2]
```diff
@@ -325,30 +326,24 @@ def measure_state_vector(
...
-        probs = np.abs(tensor) ** 2
-        probs = np.transpose(probs, indices)
-        probs = probs.reshape(-1)
+    state = state.reshape((-1,))
+    probs = np.abs(state) ** 2
+    not_measured = [i for i in range(len(qid_shape)) if i not in indices]
+    if linalg.can_numpy_support_shape(qid_shape):
+        # Use numpy transpose if we can since it's more efficient.
+        probs = probs.reshape(qid_shape)
+        probs = np.transpose(probs, list(indices) + not_measured)
```
Comment (rht) [inline cirq-core/cirq/sim/state_vector.py:335]: Looking at the usage of `np.transpose` elsewhere in the codebase, it looks like there is another place with similar usage: https://github.com/quantumlib/Cirq/blob/a95f009617bcdc613cfbca169fbedbfea1de85b7/cirq-core/cirq/sim/density_matrix_utils.py#L195-L213

Comment (NoureldinYosri) [reply cirq-core/cirq/sim/state_vector.py:335]: this is an exact copy of the code I'm modifying :disappointed: I refactored this part into a seprate method which I call from both locations

**Comment by NoureldinYosri** [comment]
@rht  I refactored the code so that both locations call the method.

I don't know why `cirq/study/resolver_test.py::test_custom_value_not_implemented` is failing but it seems to be not related to my PR since it's already failing at head
```
$ git branch
* master
$ pytest cirq/study/resolver_test.py -k test_custom_value_not_implemented
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.10.11, pytest-7.3.1, pluggy-1.0.0
Using --randomly-seed=4244270818
configfile: pyproject.toml
plugins: forked-1.6.0, cov-4.0.0, xdist-2.2.1, randomly-3.12.0, anyio-3.6.2, asyncio-0.21.0
asyncio: mode=strict
collected 47 items / 46 deselected / 1 selected                                                                                                                                                           

cirq/study/resolver_test.py x                                                                                                                                                                       [100%]

==================================================================================== 46 deselected, 1 xfailed in 0.93s ====================================================================================
```

**Comment by NoureldinYosri** [comment]
@rht this is now ready for merging

**Comment by rht** [comment]
Nit: I searched for the files named *_util.py; there are 2 occurrences. While *_utils.py has 7 occurrences. As such, I think it should be simulation_utils.py?

File: cirq-core/cirq/sim/simulation_util.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Sequence, Tuple
+
+import numpy as np
+
+from cirq import linalg
+
+
+def state_probabilities(
```
Comment (rht) [inline cirq-core/cirq/sim/simulation_util.py:21]: I find `state_probabilities_by_indices` as a function name makes it clearer that it sums out the non-measured qubits. Given that the input is already `state_probability`. What do you think?

File: cirq-core/cirq/sim/simulation_util.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    state_probability: np.ndarray, indices: Sequence[int], qid_shape: Tuple[int, ...]
+) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (NoureldinYosri) [reply cirq-core/cirq/sim/simulation_util.py]: done

File: cirq-core/cirq/sim/simulation_util.py (hunk 3 above)
Comment (NoureldinYosri) [reply cirq-core/cirq/sim/simulation_util.py]: done

**Comment by NoureldinYosri** [comment]
this is ready for another look

**Comment by tanujkhattar** [comment]
The 32 qubit dimension constraint for numpy is more annoying that I initially thought. Another example: `cirq.unitary()` fails when the dimension of circuit is > 16. Note that a 2**17 x 2 ** 17 numpy array of complex64 takes ~128gb, so it's not unreasonable to expect that the protocol should work. We can add this example as part of the issue and track it in a separate PR though. 

```
In [1]: import cirq
circuit = cirq.Cir
In [2]: q = cirq.LineQubit.range(17)
In [3]: circuit = cirq.Circuit(cirq.Z.on_each(*q))
In [4]: cirq.unitary(circuit)
```

**Comment by NoureldinYosri** [comment]
@tanujkhattar wouldn't computing such a unitary generally (when the unitary is not a tensor product) take a couple of hours/days? ...  a single multiplication of two a $2^{17} \times 2^{17}$ matricies will take around half an hour  $8^{17} \textit{operations} \approx 2251 \textit{seconds} \approx 37 \textit{minutes}$ (assumming $10^{12}$ operations per second).

---

anyway lets keep track of all the places suffering from this problem in the original issue.

---

can we merge this PR?

**Comment by NoureldinYosri** [comment]
@tanujkhattar ping

**Review by tanujkhattar — APPROVED** [review]

**Pull request merged**